qualtran/bloqs/block_encoding/linear_combination.py

```diff
@@ -0,0 +1,351 @@
+    \log_2 n \rceil + \max_i a_i, (\sum_i \lvert\lambda_i\rvert)\max_i \epsilon_i)$-block encoding
+    of $\lambda_1 U_1 + \cdots + \lambda_n U_n$.
+
+    Under the hood, this bloq uses LCU Prepare and Select oracles to build the block encoding.
+    These oracles will be automatically instantiated if not specified by the user.
+
+    Args:
+        _block_encodings: A sequence of block encodings.
+        _lambd: Corresponding coefficients.
+        lambd_bits: Number of bits needed to represent coefficients precisely.
+        _prepare: If specified, oracle preparing $\sum_i \sqrt{|\lambda_i|} \ket{i}$
+            (state should be normalized and can have junk).
+        _select: If specified, oracle taking
+            $\ket{i}\ket{\psi} \mapsto \text{sgn}(\lambda_i) \ket{i} U_i\ket{\psi}$.
+
+    Registers:
+        system: The system register.
+        ancilla: The ancilla register (present only if bitsize > 0).
+        resource: The resource register (present only if bitsize > 0).
+
+    References:
+        [Quantum algorithms: A survey of applications and end-to-end complexities](
+        https://arxiv.org/abs/2310.03011). Dalzell et al. (2023). Ch. 10.2.
+    """
+
+    _block_encodings: Tuple[BlockEncoding, ...] = field(
+        converter=lambda x: x if isinstance(x, tuple) else tuple(x), validator=validators.min_len(2)
+    )
+    _lambd: Tuple[float, ...] = field(converter=lambda x: x if isinstance(x, tuple) else tuple(x))
+    lambd_bits: SymbolicInt
+
+    _prepare: Optional[BlackBoxPrepare] = None
+    _select: Optional[BlackBoxSelect] = None
+
+    def __attrs_post_init__(self):
+        if len(self._block_encodings) != len(self._lambd):
+            raise ValueError("Must provide the same number of block encodings and coefficients.")
+        if sum(abs(x) for x in self._lambd) == 0:
+            raise ValueError("Coefficients must not sum to zero.")
+        if not all(be.system_bitsize == self.system_bitsize for be in self._block_encodings):
+            raise ValueError("All block encodings must have the same dtype.")
+        if (
+            self._prepare is not None or self._select is not None
+        ) and self.prepare.selection_registers != self.select.selection_registers:
+            raise ValueError(
+                "If given, prepare and select oracles must have same selection registers."
+            )
+        if self._select is not None and self._select.target_registers != (
+            self.signature.get_left("system"),
+        ):
+            raise ValueError(
+                "If given, select oracle must have block encoding `system` register as target."
+            )
+
+    @cached_property
+    def block_encodings(self):
+        """Appropriately negated constituent block encodings."""
+        return tuple(
+            Phase(be, phi=1, eps=0) if l < 0 else be
+            for l, be in zip(self._lambd, self._block_encodings)
```

Comment:
is this right? we ignore `be` if its coefficient is less than zero?

Reply:
Sorry, what do you mean here? If the coefficient is less than zero we use `Phase(be, phi=1, eps=0)` which negates `be`, otherwise we use `be` unmodified.

Reply:
doh I didn't see that `be` is being passed to `Phase`. I thought you were just emitting a `GlobalPhase` operation and ignoring that entry